src/System.Diagnostics.PerformanceCounter/src/System/Diagnostics/DiagnosticsConfiguration.cs

```diff
@@ -0,0 +1,105 @@
+
+        internal static int PerformanceCountersFileMappingSize
+        {
+            get
+            {
+                for (int retryCount = 0; !CanInitialize() && retryCount <= 5; ++retryCount)
+                {
+                    if (retryCount == 5)
+                        return SharedPerformanceCounter.DefaultCountersFileMappingSize;
+
+                    System.Threading.Thread.Sleep(200);
+                }
+
+                Initialize();
+                SystemDiagnosticsSection configSectionSav = s_configSection;
+                if (configSectionSav != null && configSectionSav.PerfCounters != null)
+                {
+                    int size = configSectionSav.PerfCounters.FileMappingSize;
+                    if (size < SharedPerformanceCounter.MinCountersFileMappingSize)
+                        size = SharedPerformanceCounter.MinCountersFileMappingSize;
+
+                    if (size > SharedPerformanceCounter.MaxCountersFileMappingSize)
+                        size = SharedPerformanceCounter.MaxCountersFileMappingSize;
+
+                    return size;
+                }
+                else
+                    return SharedPerformanceCounter.DefaultCountersFileMappingSize;
+            }
+        }
+
+        internal static string ConfigFilePath
+        {
+            get
+            {
+                Initialize();
+                SystemDiagnosticsSection configSectionSav = s_configSection;
+                return configSectionSav?.ElementInformation?.Source ?? string.Empty;
+            }
+        }
+
+        private static SystemDiagnosticsSection GetConfigSection()
+        {
+            SystemDiagnosticsSection configSection = (SystemDiagnosticsSection)ConfigurationManager.GetSection("system.diagnostics");
+            return configSection;
+        }
+
+        internal static bool CanInitialize()
+        {
+            bool setConfigurationSystemInProgress = (bool)(typeof(ConfigurationManager).GetProperty("SetConfigurationSystemInProgress", BindingFlags.NonPublic | BindingFlags.Static).GetValue(null));
+            return (s_initState != InitState.Initializing) && !setConfigurationSystemInProgress;
+        }
+
+        internal static void Initialize()
+        {
+            // Initialize() is also called by other components outside of Trace (such as PerformanceCounter)
+            // as a result using one lock for this critical section and another for Trace API critical sections  
+            // (such as Trace.WriteLine) could potentially lead to deadlock between 2 threads that are 
+            // executing these critical sections (and consequently obtaining the 2 locks) in the reverse order. 
+            // Using the same lock for DiagnosticsConfiguration as well as TraceInternal avoids this issue. 
+            // Sequential locks on TraceInternal.critSec by the same thread is a non issue for this critical section.
+            lock (TraceInternal.critSec)
```

Comment:
@ViktorHofer @danmosemsft I am worried about how I've used this lock. In desktop, this lock is shared between PerformanceCounter and Trace, but here they are separate and the lock is internal to TraceSource. I've essentially made a copy of that file in PerformanceCounter, but I think that is likely wrong, based on the comment (@brianrob suggested this could be the case). Is there anyway to allow PerformanceCounter to use a class internal to TraceSource so they continue to share the lock?

Reply:
Not any good way, but first, do we understand what it's trying to protect? I can see that they might want to avoid reentrancy (if bits of the configuration code end up potentially calling into perf counters eg if machine.config defines a perf counter maybe?) but the lock wouldn't prevent reentrancy. 

Mostly it is being used to avoid concurrent access to trace listeners:
https://referencesource.microsoft.com/#System/compmod/system/diagnostics/TraceInternal.cs,98190c38929a4cec,references

Perhaps `GetConfigSection()` -- probably the code being protected here, since `s_initState` isn't protected in `Refresh()` -- somehow calls into trace listeners at a lower level than the locking in trace API? 

There may possibly be value in this lock as is is here in that it prevents two threads from getting the "system.diagnostics" configuration section concurrently (at least through this class). 

@stephentoub  can you suggest why the lock was here, and why it was also used to protect Trace and the trace listeners?

Reply:
There is also a lesson here in code commenting .. if you're locking something that's not immediately evident